### PR TITLE
Make System Auth tenant id nullable

### PR DESCRIPTION
**Pull request by kjaksik**
**Description**

Changes proposed in this pull request:

- change the `tenant_id` field in `system_auths` table to be nullable
- rework the repository structure

**Related issue(s)**
See #685

### CODE REVIEW

File: components/director/internal/repo/delete.go [hunk 1]
```diff
@@ -39,22 +40,31 @@ func NewDeleterGlobal(tableName string) DeleterGlobal {
 }
 
 func (g *universalDeleter) DeleteOne(ctx context.Context, tenant string, conditions Conditions) error {
-	return g.unsafeDelete(ctx, str.Ptr(tenant), conditions, true)
+	if tenant == "" {
+		return errors.New("tenant cannot be empty")
+	}
+	conditions = append(Conditions{NewEqualCondition("tenant_id", tenant)}, conditions...)
```
Comment (crabtree) [inline components/director/internal/repo/delete.go]: As the column name please use the `tenantColumn` property value

File: components/director/internal/repo/delete.go [hunk 2]
```diff
@@ -39,22 +40,31 @@ func NewDeleterGlobal(tableName string) DeleterGlobal {
...
+	logrus.Info(conditions)
+	return g.unsafeDelete(ctx, conditions, true)
 }
 
 func (g *universalDeleter) DeleteMany(ctx context.Context, tenant string, conditions Conditions) error {
-	return g.unsafeDelete(ctx, str.Ptr(tenant), conditions, false)
+	if tenant == "" {
+		return errors.New("tenant cannot be empty")
+	}
+	conditions = append(Conditions{NewEqualCondition("tenant_id", tenant)}, conditions...)
```
Comment (crabtree) [inline components/director/internal/repo/delete.go]: As the column name please use the `tenantColumn` property value

File: components/director/internal/repo/exist.go [hunk 3]
```diff
@@ -34,14 +32,18 @@ func NewExistQuerierGlobal(tableName string) ExistQuerierGlobal {
 }
 
 func (g *universalExistQuerier) Exists(ctx context.Context, tenant string, conditions Conditions) (bool, error) {
-	return g.unsafeExists(ctx, str.Ptr(tenant), conditions)
+	if tenant == "" {
+		return false, errors.New("tenant cannot be empty")
+	}
+	conditions = append(Conditions{NewEqualCondition("tenant_id", tenant)}, conditions...)
```
Comment (crabtree) [inline components/director/internal/repo/exist.go]: As the column name please use the `tenantColumn` property value

File: components/director/internal/repo/get.go [hunk 4]
```diff
@@ -44,14 +42,18 @@ func NewSingleGetterGlobal(tableName string, selectedColumns []string) SingleGet
 }
 
 func (g *universalSingleGetter) Get(ctx context.Context, tenant string, conditions Conditions, orderByParams OrderByParams, dest interface{}) error {
-	return g.unsafeGet(ctx, str.Ptr(tenant), conditions, orderByParams, dest)
+	if tenant == "" {
+		return errors.New("tenant cannot be empty")
+	}
+	conditions = append(Conditions{NewEqualCondition("tenant_id", tenant)}, conditions...)
```
Comment (crabtree) [inline components/director/internal/repo/get.go]: As the column name please use the `tenantColumn` property value

File: components/director/internal/tenantmapping/mapper_for_system_auth.go [hunk 5]
```diff
@@ -95,12 +95,14 @@ func (m *mapperForSystemAuth) getTenantAndScopesForApplicationOrRuntime(sysAuth
...
-	if hasTenant && tenant != sysAuth.TenantID {
+	if hasTenant && tenant != *sysAuth.TenantID {
 		return "", "", errors.New("tenant missmatch")
 	}
-
-	tenant = sysAuth.TenantID
-
+	if sysAuth.TenantID != nil {
+		tenant = *sysAuth.TenantID
+	} else {
```
Comment (crabtree) [inline components/director/internal/tenantmapping/mapper_for_system_auth.go]: You can remove this `else` block as the default for the string type in go is an empty string

File: components/director/internal/tenantmapping/mapper_for_system_auth.go [hunk 6]
```diff
@@ -95,12 +95,14 @@ func (m *mapperForSystemAuth) getTenantAndScopesForApplicationOrRuntime(sysAuth
 		hasTenant = false
 	}
 
-	if hasTenant && tenant != sysAuth.TenantID {
+	if hasTenant && tenant != *sysAuth.TenantID {
```
Comment (crabtree) [inline components/director/internal/tenantmapping/mapper_for_system_auth.go]: Here may happen nil pointer dereference when the `sysAuth.TenantID` is nil

File: components/schema-migrator/migrations/202001210900_system_auth_tenant_nullable.up.sql [hunk 7]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE system_auths
```
Comment (crabtree) [inline components/schema-migrator/migrations/202001210900_system_auth_tenant_nullable.up.sql]: I think there should also be an update that updates all entries with value '00000000-0000-0000-0000-000000000000' to null value.

File: components/director/internal/domain/systemauth/service.go [hunk 8]
```diff
@@ -12,7 +12,7 @@ import (
 //go:generate mockery -name=Repository -output=automock -outpkg=automock -case=underscore
 type Repository interface {
 	Create(ctx context.Context, item model.SystemAuth) error
-	GetByID(ctx context.Context, tenant, id string) (*model.SystemAuth, error)
+	GetByID(ctx context.Context, tenant string, id string) (*model.SystemAuth, error)
```
Comment (dbadura) [inline components/director/internal/domain/systemauth/service.go]: ?

File: components/schema-migrator/migrations/202001210900_system_auth_tenant_nullable.up.sql [hunk 9]
```diff
@@ -0,0 +1,9 @@
...
+    ALTER COLUMN tenant_id DROP NOT NULL;
+
+UPDATE system_auths
+    SET tenant_id = null WHERE tenant_id = '00000000-0000-0000-0000-000000000000';
```
Comment (dbadura) [inline components/schema-migrator/migrations/202001210900_system_auth_tenant_nullable.up.sql]: UPPER CASE on NULL

**Pull request merged**